pkg/operator/controller.go

```diff
@@ -1,484 +1,331 @@
+}
 
-	pod, node struct {
-		informerEnabled bool
-		stopCh          chan struct{}
-	}
+// SetupWithManager creates a new Tuned Controller and adds it to the Manager.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func (r *TunedReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
-	pc *ProfileCalculator
-}
+	profilePredicates := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if !r.validateUpdateEvent(&e) {
+				return false
+			}
 
-type wqKey struct {
-	kind      string // object kind
-	namespace string // object namespace
-	name      string // object name
-	event     string // object event type (add/update/delete) or pass the full object on delete
-}
+			profileOld := e.ObjectOld.(*tunedv1.Profile)
+			profileNew := e.ObjectNew.(*tunedv1.Profile)
 
-func NewController() (*Controller, error) {
-	kubeconfig, err := ntoclient.GetConfig()
-	if err != nil {
-		return nil, err
+			return !reflect.DeepEqual(profileOld.Status.Conditions, profileNew.Status.Conditions)
+		},
 	}
 
-	listers := &ntoclient.Listers{}
-	clients := &ntoclient.Clients{}
-	controller := &Controller{
-		kubeconfig: kubeconfig,
-		workqueue:  workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		listers:    listers,
-		clients:    clients,
-		pc:         NewProfileCalculator(listers, clients),
+	podPredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return r.state.enablePodEvents
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return r.state.enablePodEvents
+		},
 	}
 
-	// Initial event to bootstrap CR if it doesn't exist.
-	controller.workqueue.AddRateLimited(wqKey{kind: wqKindTuned, name: tunedv1.TunedDefaultResourceName})
+	// In order to use cached client list filtering for pods: client.MatchingFields{"spec.nodeName": mNodeName}, there is a need to use and index field.
+	// The alternative is to use an uncached client.
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &corev1.Pod{}, "spec.nodeName", func(o client.Object) []string {
+		pod := o.(*corev1.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
 
-	controller.clients.Kube, err = kubeset.NewForConfig(controller.kubeconfig)
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&tunedv1.Tuned{}).
+		Owns(&tunedv1.Profile{}, builder.WithPredicates(profilePredicates)).
+		Watches(
+			&source.Kind{Type: &corev1.Node{}},
+			handler.EnqueueRequestsFromMapFunc(r.nodeToTunedRequests),
+			//TODO - consider using builder.OnlyMetadata instead these perdicats
+			builder.WithPredicates(predicate.LabelChangedPredicate{})).
+		Watches(
+			&source.Kind{Type: &corev1.Pod{}},
+			handler.EnqueueRequestsFromMapFunc(r.podLabelsToTuned),
+			builder.WithPredicates(podPredicates,
+				predicate.And(
+					predicate.Funcs{UpdateFunc: func(e event.UpdateEvent) bool {
+						return r.state.enablePodEvents
+					}}, predicate.LabelChangedPredicate{}))).
+		Watches(
```

Comment:
What about using predicates for MachineConfig pools, I think we are interested in some specific changes under the MCP.

Reply:
will be fixed in the next push